evm/src/contracts/CircleIntegration/Logic.sol

```diff
@@ -0,0 +1,227 @@
+// SPDX-License-Identifier: Apache 2
+
+pragma solidity ^0.8.22;
+
+import {IWormhole} from "src/interfaces/IWormhole.sol";
+import {IMessageTransmitter} from "src/interfaces/IMessageTransmitter.sol";
+import {ITokenMessenger} from "src/interfaces/ITokenMessenger.sol";
+import {ITokenMinter} from "src/interfaces/ITokenMinter.sol";
+import {ICircleIntegration} from "src/interfaces/ICircleIntegration.sol";
+
+import {Utils} from "src/libraries/Utils.sol";
+import {WormholeCctpMessages} from "src/libraries/WormholeCctpMessages.sol";
+
+import {Governance} from "./Governance.sol";
+import {
+    getChainToDomain,
+    getConsumedVaas,
+    getDomainToChain,
+    getInitializedImplementations,
+    getRegisteredEmitters
+} from "./Storage.sol";
+
+abstract contract Logic is ICircleIntegration, Governance {
+    using Utils for address;
+    using WormholeCctpMessages for *;
+
+    /// @inheritdoc ICircleIntegration
+    function transferTokensWithPayload(
+        TransferParameters calldata transferParams,
+        uint32 wormholeNonce,
+        bytes calldata payload
+    ) public payable returns (uint64 wormholeSequence) {
+        // Is the foreign Wormhole Circle Integration registered?
+        bytes32 targetCaller = getRegisteredEmitters()[transferParams.targetChain];
+
+        // NOTE: Reverting with Error(string) comes from the old implementation, so we preserve it.
+        require(targetCaller != 0, "target contract not registered");
+
+        // NOTE: We do not protect against re-entrancy here because we rely on the CCTP Token
+        // Messenger contract to protect against any possible re-entrancy. We are leaning on the
+        // fact that the Token Messenger keeps track of its local tokens, which are the only tokens
+        // it allows to burn.
+
+        // Approve the Token Messenger to spend tokens.
+        setTokenMessengerApproval(transferParams.token, transferParams.amount);
```

Comment:
Any reason why the approval step is here instead of inside the library `burnAndPublishMethod`?

Is it so that integrators who choose to inherit `WormholeCctp` can choose to use their own approval logic?

Reply:
Yes exactly. For `CircleIntegration`, we allow any CCTP token to invoke the token messenger's deposit for burn method. Currently the only token with CCTP enabled is USDC. But there can be more like EUROC, which is the reason why we set the approval for every outbound transfer.

But like you said, an integrator can inherit the `WormholeCctp` abstract and use his own approval logic. For example, if an integrator only cares about one token, he can set infinite approval for the CCTP Token Messenger in its constructor.